iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp

```diff
@@ -30,24 +33,44 @@ namespace roudi
 class ServiceRegistry
 {
   public:
-    static constexpr uint32_t MAX_INSTANCES_PER_SERVICE = 100u;
-    using CaproIdString_t = capro::IdString_t;
-    using InstanceSet_t = cxx::vector<CaproIdString_t, MAX_INSTANCES_PER_SERVICE>;
-    struct instance_t
+    enum class ServiceRegistryError
     {
-        InstanceSet_t instanceSet;
+        INVALID_STATE,
+        SERVICE_DESCRIPTION_ALREADY_ADDED,
+        SERVICE_REGISTRY_FULL,
     };
-    using serviceMap_t = std::map<CaproIdString_t, instance_t>;
+    /// @todo #415 should be connected with iox::MAX_NUMBER_OF_SERVICES
+    static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 100U;
+    using CaproIdString_t = capro::IdString_t;
+    using ServiceDescriptionVector_t = cxx::vector<capro::ServiceDescription, MAX_SERVICE_DESCRIPTIONS>;
+
+    /// @brief Adds given service description to registry
+    /// @param[in] serviceDescription, service to be added
+    /// @return ServiceRegistryError, error wrapped in cxx::expected
+    cxx::expected<ServiceRegistryError> add(const capro::ServiceDescription& serviceDescription);
+
+    /// @brief Removes given service description from registry
+    /// @param[in] serviceDescription, service to be removed
+    /// @return true, if service description was removed, false otherwise
+    bool remove(const capro::ServiceDescription& serviceDescription);
+
+    /// @brief Removes given service description from registry
+    /// @param[in] searchResult, reference to the vector which will be filled with the results
+    /// @param[in] service, string or wildcard to search for
+    /// @param[in] service, string or wildcard to search for
+    void find(ServiceDescriptionVector_t& searchResult,
+              const CaproIdString_t& service = capro::AnyServiceString,
+              const CaproIdString_t& instinstanceance = capro::AnyInstanceString) const;
 
-    void add(const CaproIdString_t& service, const CaproIdString_t& instance);
-    void remove(const CaproIdString_t& service, const CaproIdString_t& instance);
-    void find(InstanceSet_t& instances,
-              const CaproIdString_t& service,
-              const CaproIdString_t& instance = capro::AnyInstanceString) const;
-    const serviceMap_t& getServiceMap() const;
+    /// @brief Returns all service descriptions as copy
+    /// @return ServiceDescriptionVector_t, copy of complete service registry
+    const ServiceDescriptionVector_t getAllServices() const;
 
   private:
-    mutable serviceMap_t m_serviceMap;
+    /// @todo #859 replace std::map with prefix tree
+    ::std::multimap<CaproIdString_t, uint64_t> m_serviceMap;
```

Comment:
We cannot use a structure like this in a static memory world, it just does not work resource wise. The way it is designed in the PR it must be a multimap. 

 It will not be a multimap in the end anyway but a combination of two prefix trees and some vector to store the actual data. This will be theoretically optimal for all operations, unlike the above construct which cannot be (regardless of how it is used due to logarithmic lower bounds  of map operations for optimal implementations).
I do not know how this is supposed to work from the code alone, but I suspect there is a misunderstanding of what exactly the underlying algorithm needs to accomplish. For example we want to store full descriptor objects `ServiceDescription', not just strings which is missing in the proposal above. This is due to the fact that it can and will contain additional data describing the service in the future (e.g. for DDS).

I would also argue `fast access` and efficiency  cannot be achieved with the proposed structure. Furthermore the memory requirements if there are capacities involved for these structures are prohibitive.
Logically a solution with tries is also very concise and efficient, we just need handle the actual storage. 

I for one am not able to provide map and set implementations which are both space and time efficient at the same time given our constraints and would efficiently work here.

Reply:
@elfenpiff Please have a look at commit 47e89f10e2f916c3c3a1aca84a996096c31fa8d0, where I estimated the complexity of an approach quite similar to yours. I used `std::multimap<std::pair<CaproIdString_t, CaproIdString_t>, `ServiceDescription`>` (Attempt A).

```
find(K1,K2): O(log n+ #result)
find(K1,*) or find(*,K2): O(#entriesInMap + n^3)
find(*,*): O(#entriesInMap + #entriesInMap + n^3)

where K1 is the service string and K2 the instance string
```

On the other hand with the current approach in this PR (Attempt B):

```
find(K1,K2): O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
find(K1,*) or find(*,K2): O(log n + #result)
find(*,*): O(1)

where K1 is the service string and K2 the instance string
```

IMHO the attempt B is superior to the attempt A. Of course your approach with using a nested `set` might speed up things. However, we would still need to search with `equal_range` (O(log n)) in two maps.

> Why is it a multimap?

`multimap`'s are used as there a service or instance string can be used more than once, imagine e.g.:
|Service (1st `multimap`) | Instance (2nd `multimap`) | Event|
|----|----|-----|
|Radar|FrontLeft|Objects|
|Radar|FrontLeft|Odometry|
|Radar|FrontRight|Objects|
|Video|FrontLeft|Objects|

> For example we want to store full descriptor objects `ServiceDescription', not just strings which is missing in the proposal above. This is due to the fact that it can and will contain additional data describing the service in the future (e.g. for DDS).

That exactly, the reason. Thanks for pointing that out! For comparability with DDS it might be useful to have e.g. QoS settings attached to every `ServiceDescription.`

> I would also argue fast access and efficiency cannot be achieved with the proposed structure. 

I second that! IMHO a prefix tree is exactly what we need for efficient string look-ups. Let's have a look at the first code snippets of #859 and then do the final decision here.

Reply:
> > Why is it a multimap?

> multimap's are used as there are service or instance string that can be used more than once, imagine e.g.:

This is why one should use a `std::map<CaproIdString_t, std::map<CaproIdString_t, std::set<CaproIdString_t>>` then you have quiet an efficient layout of containers which can be reused. The more I think about it the more I doubt that the `cxx::trie` prefix tree is the right data structure for this problem.

1. We most likely will not reuse it somewhere else in iceoryx
2. But there are multiple places where a `cxx::map` can be reused
3. `cxx::trie` implementation is complex, cost a lot of effort and does not gain much performance (especially for a task like service registration which is not on the hot path)
4. A very good alternative is just two maps and a set.

Here I somehow fear we are pursuing the wrong approach. Normally we say:
- Make it work, make it beautiful and then make it fast

Here it seems we start with make it fast for a use case outside the hot path and want to pay for a very specialized data structure.

@MatthiasKillat
> The way it is designed in the PR it must be a multimap. We want to support fast wildcard searches and the individual strings do not uniquely identify the ServiceDescription.

https://github.com/eclipse-iceoryx/iceoryx/pull/860#discussion_r655486388

Reply:
@mossmaurice For our use case runtime estimation with the big O notations are the wrong approach.

Let's take a look the definition: https://en.wikipedia.org/wiki/Big_O_notation and apply it to `find(K1,K2) = O(log n+ #results)`.

This means we have: ` | find(K1, K2) | < M * (log(n) + #result)` for a real number M > 0. In our use case `n` is the number of services and lets say we grow and support a massive amount of services like `10000` (at the moment it is 512). Therefore we have `log2(10000) = 13.2877`. For 100000 it would be `log2(100000) = 16.609`.

This means the number of operations for 10000 are `|find(K1, K2)| < M * (13.2877 + #result)`.

The big o notations provides you runtime estimation for huge numbers and we are having tiny numbers here and then the constant `M` becomes very important.

Since one approach can have a constant with a value of `M = 2` and another one with `M = 1000`. Now lets use it for the runtime estimation for your two approaches with n = 100000 (number of services). For simplicity, to demonstrate my argument, I get rid of all the `#result` etc stuff.

Approach A with `M=1000`: `|find(K1, K2)| < M * (log(n)) = 1000 * log2(100000) = 1000 * 16.609 = 16609`
Approach B with `M=2`: `|find(K1, K2)| < M * (log(n) * log(n)) = 2 * (log2(100000) + log2(100000)) = 2 * (16.609 + 16.609) = 2 * 33.218 = 66.436`.

As you can see your inferior algorithm B can be much faster even the big o notation states otherwise. The big o notation states only that for `n` marching to infinity approach A is faster.


This is why premature optimization is the root of all evil since we do not know the value of `M`. The value of `M` can be high or low and this number is influenced by the CPU cache, cache locality of the data etc. things which are extremely hard to measure in a complex algorithm.

To conclude this do not optimize prematurely, use a common data structure like the `std::map` or upcoming `cxx::map` to solve this in a readable manner and then write a benchmark for optimization purposes when we need it.

Reply:
I understand O Notation and the notions of hidden constants (ignoring them for the estimate is the whole point, they are implementation dependent anyway). It is also of no use to insert specific numbers here, logarithms are small, I get this (no point in computing logarithms here). Nevertheless for these algorithms the estimates hold for small values already, so O Notation is not wrong to use here to get an estimate. The constant is important, but we will not know it regardless of approach.

Also, much of the time is used for many string comparisons, each of them being fairly expensive (number of them is logarithmic for each find call in a map).

Further: it is not necessarily `find(K1,K2)` which is expensive, but find(K1, *) since it naively leads to some quadratic complexity and improper use of maps to perform the search (iteration and such). And quadratic complexity is felt early, already for hundreds of elements with string comparisons being in play.
 
I am very concerned about the nesting of structures like map which will blow up memory consumption.

This is therefore not premature optimization, it is using the right tool for the job. Furthermore, if the abstraction is defined properly it will be very readable, the complexity is hidden in the search structure itself. 

The map with logarithmic guarantees is **hard** to implement, even more so with our static memory constraints. And a naive map will have e.g. linear search complexity which will be a problem (not felt in something like our icedelivery example) but very much so for real applications with hundreds or thousands of services, so we need to do it properly from the start.

Reply:
ad 1) This depends on what the future will bring, we cannot know. Still a useful abstraction and not a hugely complicated one.

ad 2) That is why we will also implement it, but we would need a set here to. While they are similar and will share some code, in a proper implementation (tree structure is almost the ame...) they need their own implementation and so on. Currently it is just used for introspection as far as I know, but yes it is a useful structure to have and we may be able to avoid some bad designs in the future with a map.

ad 3) The trie implementation is less complex then the map, especially with a minimalistic interface only. I have a dynamic memory prototype already (with an incomplete interface but basic operations). Hot path

ad 4) I dispute that for reasons stated, especially with static memory where all these structures need a well chosen capacity and it is constants gallore everywhere with some weird interactions like `MAXIMUM_NUMBER_OF_INSTANCES_PER_SERVICE` and the same for events and similar. We had these kind of problems in the past, only the involved containers where simpler (nested `vector`).